aws-xray-auto-instrumentation-agent-runtime/src/main/java/com/amazonaws/xray/agent/config/XRaySDKConfiguration.java

```diff
@@ -1,104 +1,322 @@
-    }
+        // Streaming threshold
+        if (agentConfiguration.getStreamingThreshold() >= 0) {
+            builder.withStreamingStrategy(new DefaultStreamingStrategy(agentConfiguration.getStreamingThreshold()));
+        } else {
+            throw new InvalidAgentConfigException("Invalid streaming threshold given in X-Ray Agent " +
+                    "configuration file: " + agentConfiguration.getStreamingThreshold());
+        }
 
-    public void setSamplingRuleFilePath(String samplingRuleFilePath) {
-        if (samplingRuleFilePath.isEmpty()) {
-            throw new IllegalArgumentException("SamplingRuleFilePath cannot be empty.");
+        // AWS Service handler manifest
+        if (agentConfiguration.getAwsServiceHandlerManifest() != null) {
+            int version = agentConfiguration.getAwsSDKVersion();
+            if (version != 1 && version != 2) {
+                throw new InvalidAgentConfigException("Invalid AWS SDK version given in X-Ray Agent configuration file: " + version);
+            } else {
+                this.awsSDKVersion = version;
+                try {
+                    this.awsServiceHandlerManifest = new File(agentConfiguration.getAwsServiceHandlerManifest()).toURI().toURL();
+                } catch (MalformedURLException e) {
+                    throw new InvalidAgentConfigException("Invalid AWS Service Handler Manifest location given in X-Ray Agent " +
+                            "configuration file: " + agentConfiguration.getAwsServiceHandlerManifest(), e);
+                }
+            }
         }
-        this.samplingRuleFilePath = samplingRuleFilePath;
+
+        // Non-configurable properties
+        SegmentContextResolverChain segmentContextResolverChain = new SegmentContextResolverChain();
+        segmentContextResolverChain.addResolver(new LambdaSegmentContextResolver());
+        segmentContextResolverChain.addResolver(new XRayTransactionContextResolver());
+        builder.withSegmentContextResolverChain(segmentContextResolverChain);
+
+        log.debug("Successfully configured the X-Ray Agent's recorder.");
+
+        AWSXRay.setGlobalRecorder(builder.build());
     }
 
-    public void setSegmentContextResolverChain(SegmentContextResolverChain segmentContextResolverChain) {
-        this.segmentContextResolverChain = segmentContextResolverChain;
+    private AgentConfiguration parseConfig(URL configFile) throws IOException {
+        Map<String, String> propertyMap = new HashMap<>();
+        JsonParser parser = factory.createParser(configFile);
+        parser.nextToken();
+
+        if (!parser.isExpectedStartObjectToken()) {
+            throw new InvalidAgentConfigException("X-Ray Agent configuration file is not valid JSON");
+        }
+
+        while (!parser.isClosed()) {
+            String field = parser.nextFieldName();
+            if (field == null) {
+                continue;  // Ignore trailing null fields
```

Comment:
What's a trailing null field? Is it something like `{"foo": "bar",null,null}`? Seems like we shouldn't ignore this broken JSON.

Reply:
the `nextFieldName` method returns a string if the current token represents a field name, and null otherwise. So at the end of the JSON block, it returns `null` for the end of object tokens. However, in a properly-formatted config file this means we've finished reading it, so I'll actually just change this to `return` instead of continuing.

Reply:
the `nextFieldName` method returns a string if the current token represents a field name, and null otherwise. So at the end of the JSON block, it returns `null` for the end of object tokens. However, in a properly-formatted config file this means we've finished reading it, so I'll actually just change this to `return` instead of continuing.